### PR TITLE
Optionally configure proxies for control plane

### PR DESCRIPTION
**Pull request by kke**
Related #856 

Unsets all environment variables ending with `_PROXY` (case insensitive) before (re)configuring the control plane. #395 unsets only `http_proxy`, `HTTP_PROXY` and `HTTPS_PROXY`, leaving `NO_PROXY` and others set.

The unsetting can be disabled by setting:

```yaml
control_plane:
  use_proxy: true
```

in cluster.yml (TODO: doc update)

Note: enabling proxy for the control plane will make kubernetes components communicate over the proxy unless you pay attention to the `NO_PROXY` setting.

### CODE REVIEW

File: lib/pharos/config_schema.rb [hunk 1]
```diff
@@ -153,6 +153,9 @@ def self.messages
         optional(:kubelet).schema do
           optional(:read_only_port).filled(:bool?)
         end
+        optional(:kubeadm).schema do
```
Comment (jakolehm) [inline lib/pharos/config_schema.rb]: IMHO this is not the right keyword for this... let's come up with something better:
- `api.use_proxy`
- `control_plane.use_proxy`
- `something_else.use_proxy`

@kke @jnummelin wdyt?

Comment (jnummelin) [reply lib/pharos/config_schema.rb]: I agree, we should not "leak" kubeadm to the config, there might be time where do not use it at all.

From the above suggestions, the best is `control_plane.use_proxy`.

Comment (jakolehm) [reply lib/pharos/config_schema.rb]: Alright, let's use `control_plane.use_proxy`.

Comment (kke) [reply lib/pharos/config_schema.rb]: True that.  Control plane sounds fitting.

**Comment by kke** [comment]
Should JoinNode reconfigure existing nodes?

**Comment by jakolehm** [comment]
> Should JoinNode reconfigure existing nodes?

Now that you asked this... I think `JoinNode` is not the right place for this. Maybe it should be part of the `ConfigureKubelet` phase (kubelet systemd unit is the key here).

**Comment by kke** [comment]
The JoinNode command will currently always get the proxy envs, I assume it will try to use them.

**Comment by kke** [comment]
Like anything that uses `kubeadm`/`kubectl` through `ssh.exec`

**Comment by jakolehm** [comment]
E2e seems to be broken.

**Comment by kke** [comment]
I think it's passing now, but just wondering should the setting be used for all the `kubectl` / `kubeadm` usages?

File: lib/pharos/phases/configure_kubelet.rb [hunk 2]
```diff
@@ -93,6 +93,15 @@ def existing_dropin
       def build_systemd_dropin
         options = []
         options << "Environment='KUBELET_EXTRA_ARGS=#{kubelet_extra_args.join(' ')}'"
+
```
Comment (jakolehm) [inline lib/pharos/phases/configure_kubelet.rb]: I think we should have specs for `build_systemd_dropin` (because it starts to be a pretty complex).

Comment (kke) [reply lib/pharos/phases/configure_kubelet.rb]: Added

**Comment by jakolehm** [comment]
> I think it's passing now, but just wondering should the setting be used for all the kubectl / kubeadm usages?

Not sure about that.. anyway if those need to be tweaked it can be a separate issue/pr (this is about the control-plane).

**Pull request merged**